File: yq.py
# coding=utf-8
import hashlib
import logging
import time
import requests

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

data = None


def get_data():
    global data
    url = "http://103.66.32.242:8005/zwfwMovePortal/interface/interfaceJson"

    timestamp = str(int(time.time()))
    print(timestamp)
    nonceHeader = "123456789abcdefg"
    signatureHeaderStr = timestamp + "23y0ufFl5YxIyGrI8hWRUZmKkvtSjLQA" + nonceHeader + timestamp
    signatureHeader = hashlib.sha256(signatureHeaderStr.encode(encoding="UTF-8")).hexdigest()

    payload = {
        "appId": "NcApplication",
        "paasHeader": "zdww",
        "timestampHeader": timestamp,
        "nonceHeader": nonceHeader,
        "signatureHeader": signatureHeader.upper(),
        "key": "3C502C97ABDA40D0A60FBEE50FAAD1DA"
    }

    signatureStr = timestamp + "fTN2pfuisxTavbTuYVSsNJHetwq5bJvCQkjjtiLM2dCratiA" + timestamp
    signature = hashlib.sha256(signatureStr.encode(encoding="UTF-8")).hexdigest()
    headers = {
        'x-wif-nonce': 'QkjjtiLM2dCratiA',
        'x-wif-paasid': 'smt-application',
        'x-wif-signature': signature.upper(),
        'x-wif-timestamp': timestamp,
    }

    response = requests.post(url, headers=headers, json=payload)

    # logging.info(' ====> %s', response.text)
    response_json = response.json()
    # print(json.dumps(response_json, sort_keys=True, indent=2, ensure_ascii=False))
    print(f"最后更新时间: {response_json['data']['end_update_time']}")
    print(f"\t高风险个数: {response_json['data']['hcount']}")
    for high in response_json['data']['highlist']:
        print(f"\t\t地区: {high['area_name']}")
        print(f"\t\t城市: {high['city']}")
        print(f"\t\t社区: {', '.join(high['communitys'])}")
        print(f"\t\t县: {high['county']}")
        print(f"\t\t省: {high['province']}")
        print(f"\n")
    print(f"\t中风险个数: {response_json['data']['mcount']}")
    for middle in response_json['data']['middlelist']:
        print(f"\t\t地区: {middle['area_name']}")
        print(f"\t\t城市: {middle['city']}")
        print(f"\t\t社区: {', '.join(middle['communitys'])}")
        print(f"\t\t县: {middle['county']}")
        print(f"\t\t省: {middle['province']}")
        print("\n")

#

File: yq.java
import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONObject;
import com.alibaba.fastjson.serializer.SerializerFeature;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * 
 
// https://www.52pojie.cn/thread-1589961-1-1.html
    pom.xml
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-web</artifactId>
      <version>2.6.4</version>
    </dependency>

    <dependency>
      <groupId>com.alibaba</groupId>
      <artifactId>fastjson</artifactId>
      <version>1.2.79</version>
    </dependency>

 */
public class QYQuery {
    private static final String url = "http://103.66.32.242:8005/zwfwMovePortal/interface/interfaceJson";
    private static final RestTemplate restTemplate = new RestTemplate();
    private static final String nonceHeader = "123456789abcdefg";

    public static void main(String[] args) throws Exception {
        String timestamp = Long.valueOf(System.currentTimeMillis() / 1000).toString();
        String signatureHeaderStr = timestamp + "23y0ufFl5YxIyGrI8hWRUZmKkvtSjLQA" + nonceHeader + timestamp;
        String signatureHeader = getSHA256StrJava(signatureHeaderStr).toUpperCase();

        JSONObject payload = new JSONObject();
        payload.put("appId", "NcApplication");
        payload.put("paasHeader", "zdww");
        payload.put("key", "3C502C97ABDA40D0A60FBEE50FAAD1DA");
        payload.put("timestampHeader", timestamp);
        payload.put("nonceHeader", nonceHeader);
        payload.put("signatureHeader", signatureHeader);

        String signatureStr = timestamp + "fTN2pfuisxTavbTuYVSsNJHetwq5bJvCQkjjtiLM2dCratiA" + timestamp;
        String signature = getSHA256StrJava(signatureStr).toUpperCase();

        System.out.println("=============== Request Body ==========");
        System.out.println(JSON.toJSONString(
                payload,
                SerializerFeature.PrettyFormat,
                SerializerFeature.WriteMapNullValue,
                SerializerFeature.WriteDateUseDateFormat
        ));

        HttpHeaders headers = new HttpHeaders();
        headers.add("x-wif-nonce", "QkjjtiLM2dCratiA");
        headers.add("x-wif-paasid", "smt-application");
        headers.add("x-wif-signature", signature);
        headers.add("x-wif-timestamp", timestamp);

        System.out.println("=============== Request headers ==========");
        System.out.println(JSON.toJSONString(
                headers,
                SerializerFeature.PrettyFormat,
                SerializerFeature.WriteMapNullValue,
                SerializerFeature.WriteDateUseDateFormat
        ));

        HttpEntity<JSONObject> formEntity = new HttpEntity<>(payload, headers);

        ResponseEntity<JSONObject> forEntity = restTemplate.postForEntity(url, formEntity, JSONObject.class);

        JSONObject object = forEntity.getBody();
        String pretty = JSON.toJSONString(
                object,
                SerializerFeature.PrettyFormat,
                SerializerFeature.WriteMapNullValue,
                SerializerFeature.WriteDateUseDateFormat
        );

        System.out.println("=============== Response JSON ==========");
        System.out.println(pretty);
    }

    /***
     * 利用java原生的摘要实现SHA256加密
     * @param str 加密后的报文
     * @return
     */
    public static String getSHA256StrJava(String str) throws NoSuchAlgorithmException {
        MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
        messageDigest.update(str.getBytes(StandardCharsets.UTF_8));
        return byte2Hex(messageDigest.digest());
    }
    /**
     * 将byte转为16进制
     * @param bytes
     * @return
     */
    private static String byte2Hex(byte[] bytes){
        StringBuilder stringBuffer = new StringBuilder();
        String temp;
        for (byte aByte : bytes) {
            temp = Integer.toHexString(aByte & 0xFF);
            if (temp.length() == 1) {
                //1得到一位的进行补0操作
                stringBuffer.append("0");
            }
            stringBuffer.append(temp);
        }
        return stringBuffer.toString();
    }
}
